File: alternatives/observer-pattern.cs
using System;
using System.Collections.Generic;

namespace EventsExampleCS
{
    class Program
    {
        public class BookSoldEventArgs : EventArgs
        {
            public string BookSold;
            public List<string> Books;
        }

        public class BookStore
        {
            public event EventHandler<BookSoldEventArgs> BookSold;

            private List<string> books;

            public BookStore()
            {
                books = new List<string>()
                {
                    "The God Delusion - Dawkins",
                    "Animal Farm - Orwell",
                    "Symposium - Plato",
                    "Ecce Homo - Nietzsche",
                    "The Trial - Kafka"
                };
            }

            public void Sell(string bookName)
            {
                if (books.Remove(bookName))
                {
                    BookSold?.Invoke(this, new BookSoldEventArgs()
                    {
                        BookSold = bookName,
                        Books = books
                    });
                }
            }
        }

        static void Main(string[] args)
        {
            var bs = new BookStore();

            bs.BookSold += (object sender, BookSoldEventArgs e) =>
            {
                Console.WriteLine($"{e.BookSold} sold, there are more {e.Books.Count} books");
            };

            bs.Sell("The God Delusion - Dawkins");
            bs.Sell("Muito Mais Que 5inco Minutos – Kefera");
            bs.Sell("Ecce Homo - Nietzsche");

            Console.ReadKey();
        }
    }
}

File: alternatives/observer-pattern.py
class Signal:
	def __init__(self):
		self.callbacks = []
	
	def connect(self, callback):
		self.callbacks.append(callback)
	
	def fire(self, *args):
		for callback in self.callbacks:
			callback(*args)

class BookStore:
	def __init__(self):
		self.book_sold = Signal()
		
		self.books = [
			"The God Delusion - Dawkins",
			"Animal Farm - Orwell",
			"Symposium - Plato",
			"Ecce Homo - Nietzsche",
			"The Trial - Kafka"
		]
		
	def sell(self, bookname):
		try:
			self.books.remove(bookname)
			self.book_sold.fire(bookname, self.books)
		except ValueError:
			pass

def logger(booksold, books):
	print("%s sold, there are more %d books" % (booksold, len(books)))

bs = BookStore()
bs.book_sold.connect(logger)
bs.sell("The God Delusion - Dawkins")
bs.sell("Muito Mais Que 5inco Minutos – Kefera")
bs.sell("Ecce Homo - Nietzsche")
